patches/@react-navigation+native+6.0.13.patch

```diff
@@ -0,0 +1,64 @@
+ export { default as useLinkProps } from './useLinkProps';
+ export { default as useLinkTo } from './useLinkTo';
+ export { default as useScrollToTop } from './useScrollToTop';
++export { default as stripKeysFromNavigationState } from './stripKeysFromNavigationState';
+ export * from '@react-navigation/core';
+ //# sourceMappingURL=index.js.map
+\ No newline at end of file
+diff --git a/node_modules/@react-navigation/native/lib/module/stripKeysFromNavigationState.js b/node_modules/@react-navigation/native/lib/module/stripKeysFromNavigationState.js
+new file mode 100644
+index 0000000..31ba628
+--- /dev/null
++++ b/node_modules/@react-navigation/native/lib/module/stripKeysFromNavigationState.js
+@@ -0,0 +1,17 @@
++// This function removes keys and some other properties from the navigation state 
++// in order to force react-navigation to regenerate those keys
++
++// keys - if we try to use the same keys it is going to break the navigation
++// stale - with stale: false navigation won't regenerate the state 
++// routeNames - to be sure that we have updated routeNames if something changed 
++
++// The argument for this function may be: NavigationState, NavigatoinRoute, PartialState<NavigationState>
++const stripKeysFromNavigationState = ({key, stale, routeNames, routes, state, ...rest}) => {
++  return {
++    ...rest, 
++    routes: routes && routes.map(stripKeysFromNavigationState),
++    state: state && stripKeysFromNavigationState(state)
++  }
++}
++
++export default stripKeysFromNavigationState;
+\ No newline at end of file
+diff --git a/node_modules/@react-navigation/native/lib/module/useLinking.js b/node_modules/@react-navigation/native/lib/module/useLinking.js
+index ee1a4a5..a83aa0d 100644
+--- a/node_modules/@react-navigation/native/lib/module/useLinking.js
++++ b/node_modules/@react-navigation/native/lib/module/useLinking.js
+@@ -1,4 +1,5 @@
+ import { findFocusedRoute, getActionFromState as getActionFromStateDefault, getPathFromState as getPathFromStateDefault, getStateFromPath as getStateFromPathDefault } from '@react-navigation/core';
++import stripKeysFromNavigationState from './stripKeysFromNavigationState';
+ import isEqual from 'fast-deep-equal';
+ import * as React from 'react';
+ import createMemoryHistory from './createMemoryHistory';
+@@ -171,7 +172,16 @@ export default function useLinking(ref, _ref) {
+       const record = history.get(index);
+ 
+       if ((record === null || record === void 0 ? void 0 : record.path) === path && record !== null && record !== void 0 && record.state) {
+-        navigation.resetRoot(record.state);
++        const rootState = navigation.getRootState();
++        // The top level key of current state and cached state may be different.
++        // That will happen if the user switched between wide and narrow view,
++        // because we are using different navigators for these views.
++        // In that case we need to regenerate the state with new keys to make navigation work 
```

Comment:
And I guess a follow up question would be...

Is it bad to propose that `react-navigation` _should_ support switching from one navigator to another and that the memory history should not necessarily need to have the exact same keys as the navigator?

I think I'm not quite understanding why the memory history's default behavior would be to break unexpectedly if these keys are mismatching. And if there's not some more formal route we could take to simply updating the memory history entries...

I don't know if either of these make sense to explore, but I was curious if either of them could work...

**Idea 1**

- We explicitly detect the navigator switch
- The memory history is rebuilt with new state keys

**Idea 2**

- We explicitly detect the navigator switch
- The keys are preserved instead of regenerated

Reply:
What we are doing here is conditionally switching the root navigator. How the error looks exactly is 
<img width="1143" alt="image" src="https://user-images.githubusercontent.com/67908363/214569969-b74376b5-2ae4-4cfd-8b3b-1cb5efe832f7.png">

So the root navigator has a different `key` and therefore the `RESET` action is not handled properly. 

My guess why `react-navigation` isn't supporting this use case is the fact that it's probably very rare to change the root navigator like that and use the same state for two different navigators.

**Idea 1**
The problem is the fact that history entries may and will have a different "shape" than our current navigation state. So it wouldn't be one to one to replace keys in the history entries with newly generated keys

**Idea 2**
I think that the same problem may exist here. (If I understand the idea correctly)

Reply:
>it's probably very rare to change the root navigator like that and use the same state for two different navigators

I hear you that this is a more advanced use-case. But, I'm still curious about this one...

>Is it bad to propose that react-navigation should support switching from one navigator to another and that the memory history should not necessarily need to have the exact same keys as the navigator?

Maybe there's some reasonable argument we can make that this _should_ be supported?

>The problem is the fact that history entries may and will have a different "shape" than our current navigation state. So it wouldn't be one to one to replace keys in the history entries with newly generated keys

Ah sorry, I'm not sure if I understood this. I'm sure I'm just missing something obvious here. Can we discuss an example where "history entries may and will have a different "shape" than our current navigation state"?

Reply:
To clarify, idea nr 1 looks like this:
- navigate a few times somewhere
- resize window
- new keys generated for the current navigation state so for example: key for the root navigator was `abc` now is `xyz`
- we modify every item in the memory history so the root key is now `xyz` in every saved state



Is that correct?

Reply:
Yes! Basically, the memory history is in "memory" so I'm wondering if we can modify or update it directly so that it references the newly generated keys for the new navigator. That way, when we go back or forward - the memory history keys match the navigation state already and no changes in `useLinking` need to happen.

Idea nr 2 is more like...

Why must the keys change at all if we switch to a new root navigator? I am sure there are some general gaps in my knowledge about what `react-navigation` uses the keys for that might explain why this is difficult to do. And it could also be that nobody ever thought about it because 99% of the apps will have one root navigator that never changes.

Reply:
I don't think we have direct access to history memory from the app level. It would need a patch anyway if we want to change states saved there

And about the "shape"

let's assume a situation like that:
```
/r/1                  -> ['chat']
/settings/about       -> ['chat', ['about]]
/search               -> ['chat', ['about'], 'search']        // go back with the browser back button
/settings/about       -> ['chat',['about]]                    // now resize
```

This is how history memory looks like now:
```
/r/1                   -> ['chat']                               // 0
/settings/about        -> ['chat', ['about]]                     // 1 <- browser history points here 
/search                -> ['chat', ['about'], 'search']          // 2
```


Our navigation state is `['chat', ['about]]`  and we are generating new keys for these routes so we have keys for `chat` and `about`.

Now we need to update all saved states in history memory
```
['chat']
['chat', ['about]]
['chat', ['about'], 'search']
```

What should be the key for the search route? We don't have it because the route wasn't present in the navigation state when we resized the window

And this state `['chat', ['about'], 'search']` would be used if we press the forward browser button

Reply:
https://user-images.githubusercontent.com/67908363/214663854-8ccd715f-c6b5-44b8-a75b-eddf454b61cc.mov

Here's how it would look like in practice

Reply:
>I don't think we have direct access to history memory from the app level. It would need a patch anyway if we want to change states saved there

Correct. I agree there could be alternative solutions. 👍 

>What should be the key for the search route? We don't have it because the route wasn't present in the navigation state when we resized the window

I think I'm missing something about what the state looks like after go "back" and pop a screen. 

When we go "back" in the example does the navigation state lose that screen (and it's key)? Or is it still retained somewhere in the navigation state? Assuming we never switch navigators, it seems like somehow the "forward" action still works correctly even though that screen is not in the navigation hierarchy anymore?

It previously sounded like the main issue is with the key of the root navigator? But now sounds like there is some problem with popped screens like the "Search" screen in our example?

Main question I am looking to get answered is something like...

What does `react-navigation` use the keys in the memory history for when reseting the root state? Do we need them at all?

Reply:
> What does react-navigation use the keys in the memory history for when reseting the root state? Do we need them at all?

This one would definitely be worth looking into it. And great conversation so far.